web3/middleware/filter.py

```diff
@@ -0,0 +1,326 @@
+import itertools
+
+from eth_utils import (
+    apply_key_map,
+    to_hex,
+    to_list,
+)
+
+from web3._utils.toolz import (
+    concat,
+    valfilter,
+)
+
+
+def segment_count(start, stop=None, step=5):
+    """Creates a segment counting generator
+
+    The generator returns tuple pairs of integers
+    that correspond to segments in the provided range.
+
+    :param start: The initial value of the counting range
+    :param stop: Optional, if provided the last value in the
+    counting range
+    :param step: Optional, the segment length. Default is 5.
+    :type start: int
+    :type stop: int
+    :return: returns a generator object
+
+
+    Example:
+
+    >>> segment_counter = segment_count(start=0, stop=10, step=3)
+    >>> next(segment_counter)
+    (0, 3)
+    >>> next(segment_counter)
+    (3, 6)
+    >>> next(segment_counter)
+    (6, 9)
+    >>> next(segment_counter) #  Remainder is also returned
+    (9, 10)
+    """
+    if stop is None:
+        return gen_unbounded_segments(start, step)
+    else:
+        return gen_bounded_segments(start, stop, step)
+
+
+def gen_unbounded_segments(start, step):
+    for segment in zip(
+            itertools.count(start, step),
+            itertools.count(start + step, step)):
+        yield segment
+
+
+def gen_bounded_segments(start, stop, step):
+    #  If the initial range is less than the step
+    #  just return (start, stop)
+    if start + step >= stop:
+        yield (start, stop)
+        return
+    for segment in zip(
+            range(start, stop - step + 1, step),
+            range(start + step, stop + 1, step)):
+        yield segment
+    else:
+        remainder = (stop - start) % step
+        #  Handle the remainder
+        if remainder:
+            yield (stop - remainder, stop)
+
+
+def block_ranges(start_block, last_block=None, step=5):
+    """Returns 2-tuple ranges describing ranges of block from start_block to last_block
+
+       Ranges do not overlap to facilitate use as ``toBlock``, ``fromBlock``
+       json-rpc arguments, which are both inclusive.
+    """
+
+    if last_block is not None and start_block > last_block:
+        raise TypeError(
+            "Incompatible start and stop arguments.",
+            "Start must be less than or equal to stop.")
+
+    return (
+        (from_block, to_block - 1)
+        for from_block, to_block
+        in segment_count(start_block, last_block + 1, step)
```

Comment:
The default value `last_block` could be `None` which blows up here.

Looking at the invocation, it looks like it's always defined, and never `None`. So maybe just drop the `None` default value in the signature, and you can drop the `last_block is not None` check above.